snuba/clusters/cluster.py

```diff
@@ -134,6 +134,31 @@ def get_batch_writer(
 ClickhouseWriterOptions = Optional[Mapping[str, Any]]
 
 
+class ConnectionCache:
```

Comment:
The maximum number of connections we store in a `ClickhousePool` is 20. After that the client waits for one to be free. 
If now we are going to reuse them we will have fewer `ClickhousePool` instances active, each one with a higher number of active connections. We need to ensure before hand that the total amount is not close to 20 otherwise clients will get stuck waiting for a connection.

Reply:
Are you talking about `settings.CLICKHOUSE_MAX_POOL_SIZE`? We have the default set to 25 and don't seem to get close to that currently. Though it may make sense to ultimately allow this to be configured per process.

Reply:
> don't seem to get close to that currently.
How did you find this out?

My concern is that, with this change, all the connections for all the clusters defined in the prod config will share the same `ClickhousePool` (now they are not, they are separate). Are we going to hit that 25 when we merge the connections for all clusters in the same `ColickhousePool` ?

Reply:
I looked at the metric `snuba.clickhouse.native.connections`. It normally hovers around 3 right now but it can spike up. Let me know if you think we need to bump it up a bit.

Reply:
I don't think that metric gives an accurate picture. When I `sum by` that metric, it goes up as high as 90. I think it needs to be tagged with a storage or something so we can see: if we dedupe the connections, and have fewer pools, will we start maxing out the connections to those pools.

Reply:
Sum by is not a meaningful aggregation to do on this metric since we might be running on hundreds of hosts (e.g. all the api, replacer, consumer, subscription processes). If you look at the max by host it's not often above 3 or 4.